mirgecom/fluid.py

```diff
@@ -0,0 +1,65 @@
+""":mod:`mirgecom.fluid` provides common utilities for fluid simulation.
+
+.. autofunction:: compute_local_velocity_gradient
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+import numpy as np  # noqa
+from meshmode.dof_array import DOFArray  # noqa
+from mirgecom.euler import ConservedVars
+
+
+def compute_local_velocity_gradient(discr, cv: ConservedVars):
+    r"""
+    Compute the cell-local gradient of fluid velocity.
+
+    Computes the cell-local gradient of fluid velocity from:
+
+    .. math::
+
+        \nabla{v_i} = \frac{1}{\rho}(\nabla(\rho{v_i})-v_i\nabla{\rho}),
+
+    where $v_i$ is ith velocity component.
+
+    Parameters
+    ----------
+    discr: grudge.eager.EagerDGDiscretization
+        the discretization to use
+    cv: mirgecom.euler.ConservedVars
+        the fluid conserved variables
+    Returns
+    -------
+    numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray`
+        representing $\partial_j{v_i}$. e.g. for 2D:
+        $\left( \begin{array}{cc}
+        \partial_{x}\mathbf{v}_{x}&\partial_{y}\mathbf{v}_{x} \\
+        \partial_{x}\mathbf{v}_{y}&\partial_{y}\mathbf{v}_{y} \end{array} \right)$
+    """
+    dim = discr.dim
+    velocity = cv.momentum/cv.mass
+    dmass = discr.grad(cv.mass)
+    dmom = np.array([discr.grad(cv.momentum[i]) for i in range(dim)], dtype=object)
```

Comment:
Incoming (potentially) dumb question :) Is there a reason we can't just take momentum, divide by density, and then compute the gradient?

Reply:
The formula I used follows the form given by Hesthaven and Warburton (section 7.5.2) and a nice feature/advantage is that we are taking the gradient using conserved quantities only, instead of a derived quantity (i.e. velocity). 

If I implement the form of the gradient they indicate, then using the conserved quantities has a distinct advantage since each gradient requires domain boundary data that we cannot have without applying the boundary conditions a-priori.  Unfortunately, Hesthaven says precious little about how to deal with the domain boundary, and what they do say borders on irrelevant for our particular needs.

Reply:
Since I also wondered about this before stumbling upon @thomasgibson's question and your reply here, it might be a good idea to add a comment in the code explaining it.